{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from intervaltree import Interval, IntervalTree\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def unquote(s):\n",
    "    if len(s) >= 2 and s[0] == '\"' and s[-1] == '\"':\n",
    "        return s[1:-1]\n",
    "    return s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "probeset_full = \"probesets/probesets_correct_coords.csv\"\n",
    "\n",
    "genecode_file = \"annotations/gencode.v26lift37.annotation.gtf\"\n",
    "\n",
    "result_file = \"probesets/annotated_probesets.csv\"\n",
    "allowed_chr = \"^chr[0-9]?[0-9]?[XYM]?[+-]$\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_intervaltrees():\n",
    "    headers = {}\n",
    "    intervaltrees = {}\n",
    "    with open(probeset_full) as f:\n",
    "        for i, line in enumerate(f):\n",
    "            line = line.rstrip().split()\n",
    "            if not headers:\n",
    "                headers = {key: value for value, key in enumerate(line)}\n",
    "            else:\n",
    "                probeset, chromosome, left, right, strand = [line[headers[i]] for i in [\"probeset_id\", \"chrom\", \"genocode_left\", \"genecode_right\", \"strand\"]]\n",
    "                key = chromosome + strand\n",
    "                tree = intervaltrees.setdefault(key, IntervalTree())\n",
    "                try:\n",
    "                    left = int(left)\n",
    "                    right = int(right)\n",
    "                    not_zero = left != 0 and right != 0\n",
    "                    wrong_order = left >= right\n",
    "                    store = not_zero\n",
    "                except ValueError:\n",
    "                    store = False\n",
    "                if store:\n",
    "                    if wrong_order:\n",
    "                        raise(ValueError(\"must not be equal {} {}!\".format(left, right)))\n",
    "                    tree[left:right] = probeset\n",
    "    return intervaltrees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "intervaltrees = read_intervaltrees()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2936520\n",
      "2936520\n",
      "2936520\n",
      "2936520\n"
     ]
    }
   ],
   "source": [
    "for i in intervaltrees[\"chr6+\"].search(168072610, 168072791, strict=True):\n",
    "    print(i.data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "possibilities = set()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter_probes = \n",
    "def filter_probes(intervaltrees, genecode_file):\n",
    "    filtered_probesets = {}\n",
    "    with open(genecode_file) as geneList:\n",
    "        currentGeneID = None\n",
    "        currentGenePieces = []\n",
    "        currentChromosome = None\n",
    "        leftGene = 0\n",
    "        rightGene = 0\n",
    "        for i, line in enumerate(geneList):\n",
    "            if line and line[0] == \"#\":\n",
    "                continue\n",
    "\n",
    "            line = line.rstrip().split(\"\\t\")\n",
    "            chromo, source, typ, left, right, _, strand, _, meta = line\n",
    "            left = int(left)\n",
    "            right = int(right)\n",
    "            meta = re.split(\";[ ]?\", meta)\n",
    "\n",
    "            processed_meta = {}\n",
    "            for m in meta[:-1]:\n",
    "                k, v = m.split()\n",
    "                processed_meta[k] = unquote(v)\n",
    "\n",
    "            transcript_type =  processed_meta.get(\"transcript_type\", \"\")\n",
    "            gene_type = processed_meta.get(\"gene_type\", \"\")\n",
    "            \n",
    "            possibilities.add((transcript_type, gene_type, typ))\n",
    "\n",
    "            protein_coding = transcript_type == \"protein_coding\" or gene_type == \"protein_coding\"\n",
    "            is_coding_or_UTR = typ in [\"CDS\", \"UTR\"]\n",
    "            key = chromo + strand\n",
    "            if protein_coding and is_coding_or_UTR and re.match(allowed_chr, key):\n",
    "                gene_name = processed_meta[\"gene_name\"]\n",
    "                for j in intervaltrees[key].search(left, right, strict=True):\n",
    "                    filtered_probesets[j.data] = gene_name\n",
    "    return filtered_probesets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "result = filter_probes(intervaltrees, genecode_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_strict_annotation(filename_in, probe_to_genename, filename_out):\n",
    "    headers = {}\n",
    "    with open(filename_out, \"w\") as f_out:\n",
    "        print(\"gene_name\", \"probeset_id\", \"seq5to3plus\", \"chrom\", \"strand\", \"genocode_left\", \"genecode_right\", \"x\", \"y\", sep=\"\\t\", file=f_out)\n",
    "        with open(filename_in) as f:\n",
    "            for i, line in enumerate(f):\n",
    "                line = line.rstrip().split()\n",
    "                if not headers:\n",
    "                    headers = {key: value for value, key in enumerate(line)}\n",
    "                else:\n",
    "                    probeset_id, seq5to3plus, chrom, genocode_left, genecode_right, strand, x, y = [line[headers[i]] for i in [\"probeset_id\", \"seq5to3plus\", \"chrom\", \"genocode_left\", \"genecode_right\", \"strand\", \"x\", \"y\"]]\n",
    "                    if probeset_id in probe_to_genename:\n",
    "                        print(probe_to_genename[probeset_id], probeset_id, seq5to3plus, chrom, strand, genocode_left, genecode_right, x, y, sep=\"\\t\", file=f_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "write_strict_annotation(probeset_full, result, result_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{('', '3prime_overlapping_ncRNA', 'gene'),\n",
       " ('', 'IG_C_gene', 'gene'),\n",
       " ('', 'IG_C_pseudogene', 'gene'),\n",
       " ('', 'IG_D_gene', 'gene'),\n",
       " ('', 'IG_J_gene', 'gene'),\n",
       " ('', 'IG_J_pseudogene', 'gene'),\n",
       " ('', 'IG_V_gene', 'gene'),\n",
       " ('', 'IG_V_pseudogene', 'gene'),\n",
       " ('', 'IG_pseudogene', 'gene'),\n",
       " ('', 'Mt_rRNA', 'gene'),\n",
       " ('', 'Mt_tRNA', 'gene'),\n",
       " ('', 'TEC', 'gene'),\n",
       " ('', 'TR_C_gene', 'gene'),\n",
       " ('', 'TR_D_gene', 'gene'),\n",
       " ('', 'TR_J_gene', 'gene'),\n",
       " ('', 'TR_J_pseudogene', 'gene'),\n",
       " ('', 'TR_V_gene', 'gene'),\n",
       " ('', 'TR_V_pseudogene', 'gene'),\n",
       " ('', 'antisense', 'gene'),\n",
       " ('', 'bidirectional_promoter_lncRNA', 'gene'),\n",
       " ('', 'lincRNA', 'gene'),\n",
       " ('', 'macro_lncRNA', 'gene'),\n",
       " ('', 'miRNA', 'gene'),\n",
       " ('', 'misc_RNA', 'gene'),\n",
       " ('', 'non_coding', 'gene'),\n",
       " ('', 'polymorphic_pseudogene', 'gene'),\n",
       " ('', 'processed_pseudogene', 'gene'),\n",
       " ('', 'processed_transcript', 'gene'),\n",
       " ('', 'protein_coding', 'gene'),\n",
       " ('', 'pseudogene', 'gene'),\n",
       " ('', 'rRNA', 'gene'),\n",
       " ('', 'scRNA', 'gene'),\n",
       " ('', 'sense_intronic', 'gene'),\n",
       " ('', 'sense_overlapping', 'gene'),\n",
       " ('', 'snRNA', 'gene'),\n",
       " ('', 'snoRNA', 'gene'),\n",
       " ('', 'transcribed_processed_pseudogene', 'gene'),\n",
       " ('', 'transcribed_unitary_pseudogene', 'gene'),\n",
       " ('', 'transcribed_unprocessed_pseudogene', 'gene'),\n",
       " ('', 'unitary_pseudogene', 'gene'),\n",
       " ('', 'unprocessed_pseudogene', 'gene'),\n",
       " ('', 'vaultRNA', 'gene'),\n",
       " ('3prime_overlapping_ncRNA', '3prime_overlapping_ncRNA', 'exon'),\n",
       " ('3prime_overlapping_ncRNA', '3prime_overlapping_ncRNA', 'transcript'),\n",
       " ('3prime_overlapping_ncRNA', 'processed_transcript', 'exon'),\n",
       " ('3prime_overlapping_ncRNA', 'processed_transcript', 'transcript'),\n",
       " ('IG_C_gene', 'IG_C_gene', 'CDS'),\n",
       " ('IG_C_gene', 'IG_C_gene', 'UTR'),\n",
       " ('IG_C_gene', 'IG_C_gene', 'exon'),\n",
       " ('IG_C_gene', 'IG_C_gene', 'stop_codon'),\n",
       " ('IG_C_gene', 'IG_C_gene', 'transcript'),\n",
       " ('IG_C_pseudogene', 'IG_C_pseudogene', 'exon'),\n",
       " ('IG_C_pseudogene', 'IG_C_pseudogene', 'transcript'),\n",
       " ('IG_D_gene', 'IG_D_gene', 'CDS'),\n",
       " ('IG_D_gene', 'IG_D_gene', 'UTR'),\n",
       " ('IG_D_gene', 'IG_D_gene', 'exon'),\n",
       " ('IG_D_gene', 'IG_D_gene', 'stop_codon'),\n",
       " ('IG_D_gene', 'IG_D_gene', 'transcript'),\n",
       " ('IG_J_gene', 'IG_J_gene', 'CDS'),\n",
       " ('IG_J_gene', 'IG_J_gene', 'UTR'),\n",
       " ('IG_J_gene', 'IG_J_gene', 'exon'),\n",
       " ('IG_J_gene', 'IG_J_gene', 'start_codon'),\n",
       " ('IG_J_gene', 'IG_J_gene', 'transcript'),\n",
       " ('IG_J_pseudogene', 'IG_J_pseudogene', 'exon'),\n",
       " ('IG_J_pseudogene', 'IG_J_pseudogene', 'transcript'),\n",
       " ('IG_V_gene', 'IG_V_gene', 'CDS'),\n",
       " ('IG_V_gene', 'IG_V_gene', 'UTR'),\n",
       " ('IG_V_gene', 'IG_V_gene', 'exon'),\n",
       " ('IG_V_gene', 'IG_V_gene', 'start_codon'),\n",
       " ('IG_V_gene', 'IG_V_gene', 'stop_codon'),\n",
       " ('IG_V_gene', 'IG_V_gene', 'transcript'),\n",
       " ('IG_V_pseudogene', 'IG_V_pseudogene', 'exon'),\n",
       " ('IG_V_pseudogene', 'IG_V_pseudogene', 'transcript'),\n",
       " ('IG_pseudogene', 'IG_pseudogene', 'exon'),\n",
       " ('IG_pseudogene', 'IG_pseudogene', 'transcript'),\n",
       " ('Mt_rRNA', 'Mt_rRNA', 'exon'),\n",
       " ('Mt_rRNA', 'Mt_rRNA', 'transcript'),\n",
       " ('Mt_tRNA', 'Mt_tRNA', 'exon'),\n",
       " ('Mt_tRNA', 'Mt_tRNA', 'transcript'),\n",
       " ('TEC', 'TEC', 'exon'),\n",
       " ('TEC', 'TEC', 'transcript'),\n",
       " ('TEC', 'processed_transcript', 'exon'),\n",
       " ('TEC', 'processed_transcript', 'transcript'),\n",
       " ('TEC', 'protein_coding', 'exon'),\n",
       " ('TEC', 'protein_coding', 'transcript'),\n",
       " ('TEC', 'transcribed_processed_pseudogene', 'exon'),\n",
       " ('TEC', 'transcribed_processed_pseudogene', 'transcript'),\n",
       " ('TEC', 'transcribed_unprocessed_pseudogene', 'exon'),\n",
       " ('TEC', 'transcribed_unprocessed_pseudogene', 'transcript'),\n",
       " ('TR_C_gene', 'TR_C_gene', 'CDS'),\n",
       " ('TR_C_gene', 'TR_C_gene', 'UTR'),\n",
       " ('TR_C_gene', 'TR_C_gene', 'exon'),\n",
       " ('TR_C_gene', 'TR_C_gene', 'stop_codon'),\n",
       " ('TR_C_gene', 'TR_C_gene', 'transcript'),\n",
       " ('TR_D_gene', 'TR_D_gene', 'CDS'),\n",
       " ('TR_D_gene', 'TR_D_gene', 'exon'),\n",
       " ('TR_D_gene', 'TR_D_gene', 'transcript'),\n",
       " ('TR_J_gene', 'TR_J_gene', 'CDS'),\n",
       " ('TR_J_gene', 'TR_J_gene', 'exon'),\n",
       " ('TR_J_gene', 'TR_J_gene', 'transcript'),\n",
       " ('TR_J_pseudogene', 'TR_J_pseudogene', 'exon'),\n",
       " ('TR_J_pseudogene', 'TR_J_pseudogene', 'transcript'),\n",
       " ('TR_V_gene', 'TR_V_gene', 'CDS'),\n",
       " ('TR_V_gene', 'TR_V_gene', 'UTR'),\n",
       " ('TR_V_gene', 'TR_V_gene', 'exon'),\n",
       " ('TR_V_gene', 'TR_V_gene', 'start_codon'),\n",
       " ('TR_V_gene', 'TR_V_gene', 'transcript'),\n",
       " ('TR_V_pseudogene', 'TR_V_pseudogene', 'exon'),\n",
       " ('TR_V_pseudogene', 'TR_V_pseudogene', 'transcript'),\n",
       " ('antisense', 'antisense', 'exon'),\n",
       " ('antisense', 'antisense', 'transcript'),\n",
       " ('antisense', 'processed_transcript', 'exon'),\n",
       " ('antisense', 'processed_transcript', 'transcript'),\n",
       " ('antisense', 'protein_coding', 'exon'),\n",
       " ('antisense', 'protein_coding', 'transcript'),\n",
       " ('bidirectional_promoter_lncRNA', 'bidirectional_promoter_lncRNA', 'exon'),\n",
       " ('bidirectional_promoter_lncRNA',\n",
       "  'bidirectional_promoter_lncRNA',\n",
       "  'transcript'),\n",
       " ('bidirectional_promoter_lncRNA', 'processed_transcript', 'exon'),\n",
       " ('bidirectional_promoter_lncRNA', 'processed_transcript', 'transcript'),\n",
       " ('lincRNA', 'lincRNA', 'exon'),\n",
       " ('lincRNA', 'lincRNA', 'transcript'),\n",
       " ('lincRNA', 'processed_transcript', 'exon'),\n",
       " ('lincRNA', 'processed_transcript', 'transcript'),\n",
       " ('macro_lncRNA', 'macro_lncRNA', 'exon'),\n",
       " ('macro_lncRNA', 'macro_lncRNA', 'transcript'),\n",
       " ('miRNA', 'lincRNA', 'exon'),\n",
       " ('miRNA', 'lincRNA', 'transcript'),\n",
       " ('miRNA', 'miRNA', 'exon'),\n",
       " ('miRNA', 'miRNA', 'transcript'),\n",
       " ('misc_RNA', 'antisense', 'exon'),\n",
       " ('misc_RNA', 'antisense', 'transcript'),\n",
       " ('misc_RNA', 'lincRNA', 'exon'),\n",
       " ('misc_RNA', 'lincRNA', 'transcript'),\n",
       " ('misc_RNA', 'misc_RNA', 'exon'),\n",
       " ('misc_RNA', 'misc_RNA', 'transcript'),\n",
       " ('misc_RNA', 'processed_transcript', 'exon'),\n",
       " ('misc_RNA', 'processed_transcript', 'transcript'),\n",
       " ('misc_RNA', 'protein_coding', 'exon'),\n",
       " ('misc_RNA', 'protein_coding', 'transcript'),\n",
       " ('misc_RNA', 'sense_overlapping', 'exon'),\n",
       " ('misc_RNA', 'sense_overlapping', 'transcript'),\n",
       " ('non_coding', 'non_coding', 'exon'),\n",
       " ('non_coding', 'non_coding', 'transcript'),\n",
       " ('non_stop_decay', 'protein_coding', 'CDS'),\n",
       " ('non_stop_decay', 'protein_coding', 'Selenocysteine'),\n",
       " ('non_stop_decay', 'protein_coding', 'UTR'),\n",
       " ('non_stop_decay', 'protein_coding', 'exon'),\n",
       " ('non_stop_decay', 'protein_coding', 'start_codon'),\n",
       " ('non_stop_decay', 'protein_coding', 'stop_codon'),\n",
       " ('non_stop_decay', 'protein_coding', 'transcript'),\n",
       " ('nonsense_mediated_decay', 'polymorphic_pseudogene', 'CDS'),\n",
       " ('nonsense_mediated_decay', 'polymorphic_pseudogene', 'UTR'),\n",
       " ('nonsense_mediated_decay', 'polymorphic_pseudogene', 'exon'),\n",
       " ('nonsense_mediated_decay', 'polymorphic_pseudogene', 'start_codon'),\n",
       " ('nonsense_mediated_decay', 'polymorphic_pseudogene', 'stop_codon'),\n",
       " ('nonsense_mediated_decay', 'polymorphic_pseudogene', 'transcript'),\n",
       " ('nonsense_mediated_decay', 'protein_coding', 'CDS'),\n",
       " ('nonsense_mediated_decay', 'protein_coding', 'Selenocysteine'),\n",
       " ('nonsense_mediated_decay', 'protein_coding', 'UTR'),\n",
       " ('nonsense_mediated_decay', 'protein_coding', 'exon'),\n",
       " ('nonsense_mediated_decay', 'protein_coding', 'start_codon'),\n",
       " ('nonsense_mediated_decay', 'protein_coding', 'stop_codon'),\n",
       " ('nonsense_mediated_decay', 'protein_coding', 'transcript'),\n",
       " ('polymorphic_pseudogene', 'polymorphic_pseudogene', 'CDS'),\n",
       " ('polymorphic_pseudogene', 'polymorphic_pseudogene', 'UTR'),\n",
       " ('polymorphic_pseudogene', 'polymorphic_pseudogene', 'exon'),\n",
       " ('polymorphic_pseudogene', 'polymorphic_pseudogene', 'start_codon'),\n",
       " ('polymorphic_pseudogene', 'polymorphic_pseudogene', 'stop_codon'),\n",
       " ('polymorphic_pseudogene', 'polymorphic_pseudogene', 'transcript'),\n",
       " ('processed_pseudogene', 'processed_pseudogene', 'exon'),\n",
       " ('processed_pseudogene', 'processed_pseudogene', 'transcript'),\n",
       " ('processed_pseudogene', 'pseudogene', 'exon'),\n",
       " ('processed_pseudogene', 'pseudogene', 'transcript'),\n",
       " ('processed_transcript', 'polymorphic_pseudogene', 'exon'),\n",
       " ('processed_transcript', 'polymorphic_pseudogene', 'transcript'),\n",
       " ('processed_transcript', 'processed_transcript', 'exon'),\n",
       " ('processed_transcript', 'processed_transcript', 'transcript'),\n",
       " ('processed_transcript', 'protein_coding', 'exon'),\n",
       " ('processed_transcript', 'protein_coding', 'transcript'),\n",
       " ('processed_transcript', 'pseudogene', 'exon'),\n",
       " ('processed_transcript', 'pseudogene', 'transcript'),\n",
       " ('processed_transcript', 'transcribed_processed_pseudogene', 'exon'),\n",
       " ('processed_transcript', 'transcribed_processed_pseudogene', 'transcript'),\n",
       " ('processed_transcript', 'transcribed_unitary_pseudogene', 'exon'),\n",
       " ('processed_transcript', 'transcribed_unitary_pseudogene', 'transcript'),\n",
       " ('processed_transcript', 'transcribed_unprocessed_pseudogene', 'exon'),\n",
       " ('processed_transcript', 'transcribed_unprocessed_pseudogene', 'transcript'),\n",
       " ('processed_transcript', 'unitary_pseudogene', 'exon'),\n",
       " ('processed_transcript', 'unitary_pseudogene', 'transcript'),\n",
       " ('protein_coding', 'IG_C_gene', 'CDS'),\n",
       " ('protein_coding', 'IG_C_gene', 'UTR'),\n",
       " ('protein_coding', 'IG_C_gene', 'exon'),\n",
       " ('protein_coding', 'IG_C_gene', 'stop_codon'),\n",
       " ('protein_coding', 'IG_C_gene', 'transcript'),\n",
       " ('protein_coding', 'IG_V_gene', 'CDS'),\n",
       " ('protein_coding', 'IG_V_gene', 'UTR'),\n",
       " ('protein_coding', 'IG_V_gene', 'exon'),\n",
       " ('protein_coding', 'IG_V_gene', 'start_codon'),\n",
       " ('protein_coding', 'IG_V_gene', 'transcript'),\n",
       " ('protein_coding', 'TR_V_gene', 'CDS'),\n",
       " ('protein_coding', 'TR_V_gene', 'exon'),\n",
       " ('protein_coding', 'TR_V_gene', 'transcript'),\n",
       " ('protein_coding', 'polymorphic_pseudogene', 'CDS'),\n",
       " ('protein_coding', 'polymorphic_pseudogene', 'UTR'),\n",
       " ('protein_coding', 'polymorphic_pseudogene', 'exon'),\n",
       " ('protein_coding', 'polymorphic_pseudogene', 'start_codon'),\n",
       " ('protein_coding', 'polymorphic_pseudogene', 'stop_codon'),\n",
       " ('protein_coding', 'polymorphic_pseudogene', 'transcript'),\n",
       " ('protein_coding', 'protein_coding', 'CDS'),\n",
       " ('protein_coding', 'protein_coding', 'Selenocysteine'),\n",
       " ('protein_coding', 'protein_coding', 'UTR'),\n",
       " ('protein_coding', 'protein_coding', 'exon'),\n",
       " ('protein_coding', 'protein_coding', 'start_codon'),\n",
       " ('protein_coding', 'protein_coding', 'stop_codon'),\n",
       " ('protein_coding', 'protein_coding', 'transcript'),\n",
       " ('pseudogene', 'TEC', 'exon'),\n",
       " ('pseudogene', 'TEC', 'transcript'),\n",
       " ('pseudogene', 'antisense', 'exon'),\n",
       " ('pseudogene', 'antisense', 'transcript'),\n",
       " ('pseudogene', 'lincRNA', 'exon'),\n",
       " ('pseudogene', 'lincRNA', 'transcript'),\n",
       " ('pseudogene', 'pseudogene', 'exon'),\n",
       " ('pseudogene', 'pseudogene', 'transcript'),\n",
       " ('pseudogene', 'sense_intronic', 'exon'),\n",
       " ('pseudogene', 'sense_intronic', 'transcript'),\n",
       " ('pseudogene', 'sense_overlapping', 'exon'),\n",
       " ('pseudogene', 'sense_overlapping', 'transcript'),\n",
       " ('rRNA', 'rRNA', 'exon'),\n",
       " ('rRNA', 'rRNA', 'transcript'),\n",
       " ('retained_intron', 'antisense', 'exon'),\n",
       " ('retained_intron', 'antisense', 'transcript'),\n",
       " ('retained_intron', 'lincRNA', 'exon'),\n",
       " ('retained_intron', 'lincRNA', 'transcript'),\n",
       " ('retained_intron', 'polymorphic_pseudogene', 'exon'),\n",
       " ('retained_intron', 'polymorphic_pseudogene', 'transcript'),\n",
       " ('retained_intron', 'processed_transcript', 'exon'),\n",
       " ('retained_intron', 'processed_transcript', 'transcript'),\n",
       " ('retained_intron', 'protein_coding', 'exon'),\n",
       " ('retained_intron', 'protein_coding', 'transcript'),\n",
       " ('retained_intron', 'pseudogene', 'exon'),\n",
       " ('retained_intron', 'pseudogene', 'transcript'),\n",
       " ('retained_intron', 'sense_intronic', 'exon'),\n",
       " ('retained_intron', 'sense_intronic', 'transcript'),\n",
       " ('retained_intron', 'transcribed_processed_pseudogene', 'exon'),\n",
       " ('retained_intron', 'transcribed_processed_pseudogene', 'transcript'),\n",
       " ('retained_intron', 'transcribed_unitary_pseudogene', 'exon'),\n",
       " ('retained_intron', 'transcribed_unitary_pseudogene', 'transcript'),\n",
       " ('retained_intron', 'transcribed_unprocessed_pseudogene', 'exon'),\n",
       " ('retained_intron', 'transcribed_unprocessed_pseudogene', 'transcript'),\n",
       " ('scRNA', 'scRNA', 'exon'),\n",
       " ('scRNA', 'scRNA', 'transcript'),\n",
       " ('sense_intronic', 'processed_transcript', 'exon'),\n",
       " ('sense_intronic', 'processed_transcript', 'transcript'),\n",
       " ('sense_intronic', 'sense_intronic', 'exon'),\n",
       " ('sense_intronic', 'sense_intronic', 'transcript'),\n",
       " ('sense_overlapping', 'processed_transcript', 'exon'),\n",
       " ('sense_overlapping', 'processed_transcript', 'transcript'),\n",
       " ('sense_overlapping', 'sense_overlapping', 'exon'),\n",
       " ('sense_overlapping', 'sense_overlapping', 'transcript'),\n",
       " ('snRNA', 'snRNA', 'exon'),\n",
       " ('snRNA', 'snRNA', 'transcript'),\n",
       " ('snoRNA', 'lincRNA', 'exon'),\n",
       " ('snoRNA', 'lincRNA', 'transcript'),\n",
       " ('snoRNA', 'snoRNA', 'exon'),\n",
       " ('snoRNA', 'snoRNA', 'transcript'),\n",
       " ('transcribed_processed_pseudogene', 'pseudogene', 'exon'),\n",
       " ('transcribed_processed_pseudogene', 'pseudogene', 'transcript'),\n",
       " ('transcribed_processed_pseudogene',\n",
       "  'transcribed_processed_pseudogene',\n",
       "  'exon'),\n",
       " ('transcribed_processed_pseudogene',\n",
       "  'transcribed_processed_pseudogene',\n",
       "  'transcript'),\n",
       " ('transcribed_unitary_pseudogene', 'protein_coding', 'exon'),\n",
       " ('transcribed_unitary_pseudogene', 'protein_coding', 'transcript'),\n",
       " ('transcribed_unitary_pseudogene', 'transcribed_unitary_pseudogene', 'exon'),\n",
       " ('transcribed_unitary_pseudogene',\n",
       "  'transcribed_unitary_pseudogene',\n",
       "  'transcript'),\n",
       " ('transcribed_unprocessed_pseudogene', 'protein_coding', 'exon'),\n",
       " ('transcribed_unprocessed_pseudogene', 'protein_coding', 'transcript'),\n",
       " ('transcribed_unprocessed_pseudogene', 'pseudogene', 'exon'),\n",
       " ('transcribed_unprocessed_pseudogene', 'pseudogene', 'transcript'),\n",
       " ('transcribed_unprocessed_pseudogene',\n",
       "  'transcribed_unprocessed_pseudogene',\n",
       "  'exon'),\n",
       " ('transcribed_unprocessed_pseudogene',\n",
       "  'transcribed_unprocessed_pseudogene',\n",
       "  'transcript'),\n",
       " ('unitary_pseudogene', 'pseudogene', 'exon'),\n",
       " ('unitary_pseudogene', 'pseudogene', 'transcript'),\n",
       " ('unitary_pseudogene', 'unitary_pseudogene', 'exon'),\n",
       " ('unitary_pseudogene', 'unitary_pseudogene', 'transcript'),\n",
       " ('unprocessed_pseudogene', 'pseudogene', 'exon'),\n",
       " ('unprocessed_pseudogene', 'pseudogene', 'transcript'),\n",
       " ('unprocessed_pseudogene', 'unprocessed_pseudogene', 'exon'),\n",
       " ('unprocessed_pseudogene', 'unprocessed_pseudogene', 'transcript'),\n",
       " ('vaultRNA', 'vaultRNA', 'exon'),\n",
       " ('vaultRNA', 'vaultRNA', 'transcript')}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "possibilities"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
